packages/core/js-client/src/compilerSupport/conversions.ts

```diff
@@ -14,222 +14,241 @@
  * limitations under the License.
  */
 
-// TODO: This file is a mess. Need to refactor it later
-/* eslint-disable */
-// @ts-nocheck
-
-import assert from "assert";
-
-import type {
+import {
   ArrowType,
   ArrowWithoutCallbacks,
-  JSONArray,
   JSONValue,
-  NonArrowType,
+  LabeledProductType,
+  NonArrowSimpleType,
+  ScalarType,
+  SimpleTypes,
+  UnlabeledProductType,
 } from "@fluencelabs/interfaces";
-import { match } from "ts-pattern";
 
-import { CallServiceData } from "../jsServiceHost/interfaces.js";
-import { jsonify } from "../util/utils.js";
+import { ParticleContext } from "../jsServiceHost/interfaces.js";
+
+import { ServiceImpl } from "./types.js";
+
+export class SchemaValidationError extends Error {
+  constructor(
+    public path: string[],
+    schema: NonArrowSimpleType | ArrowWithoutCallbacks,
+    expected: string,
+    provided: JSONValue | ServiceImpl[string],
+  ) {
+    const given =
+      provided === null
+        ? "null"
+        : Array.isArray(provided)
+        ? "array"
+        : typeof provided;
+
+    const message = `Type mismatch. Path: ${path.join(
+      ".",
+    )}; Expected: ${expected}; Given: ${given};\n\nschema: ${JSON.stringify(
+      schema,
+    )}`;
+
+    super(message);
+  }
+}
+
+interface ValidationContext {
+  path: string[];
+}
+
+const numberTypes = [
+  "u8",
+  "u16",
+  "u32",
+  "u64",
+  "i8",
+  "i16",
+  "i32",
+  "i64",
+  "f32",
+  "f64",
+] as const;
+
+function isScalar(
+  schema: ScalarType,
+  arg: JSONValue,
+  { path }: ValidationContext,
+) {
+  if (numberTypes.includes(schema.name)) {
+    if (typeof arg !== "number") {
+      throw new SchemaValidationError(path, schema, "number", arg);
+    }
+  } else if (schema.name === "bool") {
+    if (typeof arg !== "boolean") {
+      throw new SchemaValidationError(path, schema, "boolean", arg);
+    }
+  } else if (schema.name === "string") {
+    if (typeof arg !== "string") {
+      throw new SchemaValidationError(path, schema, "string", arg);
+    }
+  } else {
+    throw new SchemaValidationError(path, schema, "never", arg);
```

Comment:
maybe would be more useful
```suggestion
    throw new SchemaValidationError(path, schema, schema.name, arg);
```

Reply:
or maybe even more useful would be to throw a more actioonable erro that would explain that this is unsupported please update js-client